File: 300. Longest Increasing Subsequence/300.cpp
// dynamic programming
// time: O(n^2)
// space: O(n)
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int ans = 1;
        // stores the length of longest increasing subsequence ending at index i
        // initialize with 1 each one is it's own increasing subsequence
        vector<int> maxlen(nums.size(), 1);

        for(int i =0; i< nums.size(); i++){
            int curnum = nums[i];

            // trace back to find longest existing subsequence that ends at a value smaller than curnum
            for(int j = i-1; j>=0; j--){
                if (nums[j] < curnum){
                    maxlen[i] = max(maxlen[i], maxlen[j] + 1);
                }
            }
        }

        // find the longest length
        for(int len: maxlen){
            ans = max(ans, len);
        }

        return ans;
    }
};

File: 300. Longest Increasing Subsequence/300.py
"""
dynamic programming

"""
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if len(nums) ==0:
            return 0

        dp = [1] * len(nums)

        for i in range(1, len(nums)):
            for j in range(0, i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[j] + 1, dp[i])

        return max(dp)
"""
dynamic programming
dp[i] = length fo longest strictly increasing subsequence that ends in nums[i]

initialization:
dp[0] = 1
i should be range of indices for nums

for i from 1 to len(nums):
    dp[i] = length of longest strictly increasing subsequence that ended in j < i  + 1 where nums[j] < nums[i]
    for j from 0 to i:
        if nums[j] < nums[i]
            dp[i] = max(dp[j] + 1, dp[i])

return max(dp)
"""
